### ISSUE TITLE
Differentiable forward solver if we plan to use Torch/Tensorflow?

### ISSUE THREAD

**Issue by atulag0711** [BAMresearch/probeye#26]
Most of the ML based probabilistic packages (Pyro--> torch, Tensorflow Probability --> tensorflow) have the prerequisite that everything is differentiable. If our forward solver doesnt pass jacobians then we break the computational graph and we dont capitalize of the strengths of ML libraries. Since we have  Pyro (Torch) now and later plan to have tensorflow wrapper I think Its important to discuss this issue.

- For legacy solvers, jacobians are hard to come by, but if we plan to develop a new solver, gradients can be introduced. If we have gradients, torch autodiff can be overloaded (https://pytorch.org/docs/stable/notes/extending.html) to include the gradients thus keeping the computational graph intact. 

- If the solver is FEnics based (which I assume would be), the jacobian operator can be returned to my knowledge. (http://www.dolfin-adjoint.org/en/latest/)

- If solver is expensive, we can learn a ML model which acts as a surrogate. If its a ML model, for sure it is differntiable, so we dont have a problem.

- For a simplex solver (if for some reason its not differntiable), if anylytical gradient is known, still the backward call of autodiff can be overloaded to include the gradient.

If we tackle a very complex problem for which differential solver cant be written then:

1. Already implemented EMCEE in probEye can be used (However I dont know its strengeths and weaknesses).
2. A ML based surrogate can be build (If its complex for sure it will be expensive to call again and again for EMCEE also so this makes sense) and this surrogate can be used in the Pyro framework.
3. If people want to use Torch without a surrogate, then I am trying write a Random Walk kernel for Pyro (IMO will be useless for real world cases as forward solbver will be expensive and samples needed will bve huge for Random walk). I will also be trying to implement Black Box Variational Inference (grad free VI method).
4. More classical grad free sampling methods can be implemented. I would propose SMC.

Please provide your views and how we should move forward.

**Comment by joergfunger** [comment]
I would not reimplement solvers in pyro, but then rather use other libraries that are already existing to do so. If we realize that pyro is only interesting for problems where a gradient of the forward model is available, then let's restrict ourselves to these kind of problems in otherwise use emcee, dynesty or others. For the first simple use case, I would use an analytical function where the gradient is analytically provided. And I also agree that we might have to look at some point into response surfaces, as you have mentioned this would then allow gradients to be computed as well. In a nutshell, I think it would be nice to try to overwrite the gradient method and see how far we can get.

**Comment by TTitscher** [comment]
The analytical variational Bayes solver requires the derivative of the term "model(theta) - data" w.r.t. theta, so basically **d**model/**d**theta. So that feature is IMO generally useful -- not only for torch/tensorflow. Other examples could be deterministic least square fits, maybe for some debug purposes, that also benefit from this derivative. Or in post-processing, where the Jacobian can be used to [propagate](https://en.wikipedia.org/wiki/Propagation_of_uncertainty#Non-linear_combinations) the parameter uncertainty to the uncertainty of the model response.

Practically, I faced situations where I could provide the derivative w.r.t. a subset of all parameters analytically and used central differences for (a few) others. So something like (dummy syntax)
~~~py
def jacobian(self, prm):  
    forward_model = self
    jac = probeye.util.cdf(forward_model, prm.subset(["A", "B"])
    jac["C"] = 3 * forward_model(prm) * prm["C"]**2
    return
~~~

**Comment by joergfunger** [comment]
I think what is important is to understand what exactly we need (and I don't know myself). I think that with the above we need the derivative of all model outputs with respect to all model parameters. For a large number of outputs this requires to solve as many adjoint problems as we have outputs (even for each entry in a vector), unless we are able to write it analytically. However, the NUTS solver seems to be satisfied with dLogLike/dtheta, which even for a large number of model outputs just a single adjoint. You can certainly deduce from dmodel/dtheta dLoglike/dtheta, but with a significant overhead. In addition to the computational overhead, the intermediate matrix dModel/dtheta might be a relatively large matrix, in the second case there is always a vector returned.

**Comment by TTitscher** [comment]
dLogLike/dtheta could be an option, but I have no experience there. At first glance, calculating dLoglike/dtheta in the forward model means that it needs to know about _data/"experiment"_ and a _noise model_. In probeye, those are currently separated.

Anyways, since Atul also mentioned "Jacobian" several times, I only thought about dmodel/dtheta. How about having dmodel/dtheta as kind of a default case (worst default case via central differences of the forward model), especially since it can be provided by a surrogate model and is mathematically straight forward to "chain rule" into a dLogLike/dtheta in each noise model?
